### PR TITLE
Refactor stress components

### PR DESCRIPTION
**Pull request by Thomas-Ulrich**
Currently, there are 3 stress components in SeisSol, with partial redundancy.

- STRESS_STR_DIP_SLIP_AM (used in Sumatra SC. the optimally oriented fault loading is limited to rake=0 or 90 degree), 
- OptimalStress (used in Palu, is a generalization of STRESS_STR_DIP_SLIP_AM because we can specify the rake of the maximum traction. It also features other small differences (1) returned tensor not normalized, 2) TR(sij) used to constrain the magnitude of the stress tensor, and not s_zz, (which is not necessarily more realistic)).
- AndresonianStress (simpler parameterization for an Andersonian stress, that is all principal stresses vertical or horizontal).

At the end, so far, I kept STRESS_STR_DIP_SLIP_AM for backward compatibility but then users don't know which component to use. Therefore, we could either continue keeping STRESS_STR_DIP_SLIP_AM but insist on its deprecations (see the current state of PR), or drop STRESS_STR_DIP_SLIP_AM.

Also, I would add an extra parameter s_zz in OptimalStress ( to offer the possibility to constrain the magnitude of the stress tensor with s_zz, which at the end seems to make more sense in most cases).
This would lead to some of our previous setups needing slight updates in the easi files.
(e.g. adding s_zz 0 in OptimalStress, changing STRESS_STR_DIP_SLIP_AM to OptimalStress following the example in the readthedocs). 

I hope the aim of the PR is clear, let me know what you think,

Thomas.

### CODE REVIEW

File: doc/maps.rst [hunk 1]
```diff
@@ -234,21 +236,22 @@ magnitudes, following the procedure described in Ulrich et al.
...
 -----------------
 
-This function allows computing Andersonian stresses (for which one principal axis of the stress tensor is vertical).
+This component allows computing Andersonian stress tensors (for which one principal axis of the stress tensor is vertical).
 
 The principal stress orientations are defined by SH_max (measured from North, positive eastwards), the direction of maximum horizontal compressive stress.
 
 S_v defines which of the principal stresses :math:`s_i` is vertical where :math:`s_1>s_2>s_3`.
-S_v = 1, 2 or 3 should be used if the vertical principal stress is the maximum, intermediate or minimum compressive stress.
+S_v = 1, 2, or 3 should be used when the vertical principal stress is the maximum, intermediate, or minimum compressive stress, respetively.
```
Comment (uphoffc) [inline doc/maps.rst:254]: typo: respetively -> respectively

File: include/easi/component/OptimalStress.h [hunk 2]
```diff
@@ -45,7 +45,7 @@
 namespace easi {
   struct OptimalStress {
     struct in {
-      double mu_d, mu_s, strike, dip, rake, effectiveConfiningStress, cohesion, R, s2ratio;
+      double mu_d, mu_s, strike, dip, rake, effectiveConfiningStress, s_zz, cohesion, R, s2ratio;
```
Comment (uphoffc) [inline include/easi/component/OptimalStress.h:48]: This breaks backward compatibility, right?

File: include/easi/component/OptimalStress.h [hunk 3]
```diff
@@ -99,7 +100,16 @@ void easi::OptimalStress::evaluate() {
   o.b_yy = -((ci * ci * s11 * sr * sr + s33 * si * si) * cd * cd + 2 * ci * sd * si * sr * (s33 - s11) * cd + (s33 * sd * sd * sr * sr + cr * cr * s22) * ci * ci + sd * sd * si * si * s11) * ss * ss - 2 * cr * cs * (cd * cd * ci * sr * s11 + sd * si * (s33 - s11) * cd - ci * sr * (-s33 * sd * sd + s22)) * ss - cs * cs * (cd * cd * cr * cr * s11 + cr * cr * s33 * sd * sd + s22 * sr * sr);
   o.b_yz = -cd * sd * sr * ss * (s33 - s11) * si * si - ((-ss * (-s11 * sr * sr + s33) * ci + cr * cs * sr * s11) * cd * cd + ss * ((s33 * sr * sr - s11) * sd * sd + cr * cr * s22) * ci - cr * cs * sr * (-s33 * sd * sd + s22)) * si + cd * ci * sd * (ci * sr * ss + cr * cs) * (s33 - s11);
   o.b_zz = -(cd * cd * s11 * sr * sr + s33 * sd * sd * sr * sr + cr * cr * s22) * si * si - 2 * cd * ci * sd * sr * (s11 - s33) * si - ci * ci * (cd * cd * s33 + s11 * sd * sd);
```
Comment (uphoffc) [inline include/easi/component/OptimalStress.h:102]: This is only valid for an enu system, right? Can we mention that in the documentation?

File: include/easi/component/OptimalStress.h [hunk 4]
```diff
@@ -99,7 +100,16 @@ void easi::OptimalStress::evaluate() {
...
+  
+  if (abs(i.s_zz)>0.0) {
+    double factor = -effectiveConfiningStress/o.b_zz;
+    o.b_xx = o.b_xx * factor;  
+    o.b_yy = o.b_yy * factor;  
+    o.b_zz = o.b_zz * factor;  
+    o.b_xy = o.b_xy * factor;  
+    o.b_xz = o.b_xz * factor;  
+    o.b_yz = o.b_yz * factor;  
+  }
```
Comment (uphoffc) [inline include/easi/component/OptimalStress.h:112]: Why do you introduce s_zz as new parameter? The same functionality could be achieved by adding an AffineMap or FunctionMap?

**Review by uphoffc — CHANGES_REQUESTED** [review]
Not so sure whether I like the addition of s_zz. 
1. Breaks backward compatibility.
2. Not straightforward to see whether s_zz or effectiveConfiningStress is used.
3. It's not really necessary, could be implemented with existing components.

File: include/easi/component/OptimalStress.h (hunk 4 above)
Comment (Thomas-Ulrich) [reply include/easi/component/OptimalStress.h:112]: My point is that it makes more sense in most cases (all?) to constrain sig_zz than Tr(s_ii)/3 to be equal to the effective weight of the overlying column.


An extreme case when that makes a difference is for a stress setup favoring normal and reverse faulting:
image
Mohr–Coulomb diagram at 5 km depth for normal and reverse faults illustrated in blue and red, respectively.
Source: https://doi-org.emedien.ub.uni-muenchen.de/10.1093/gji/ggy407

normal faulting: s1 = s_zz
reverse faulting s3 = s_zz.
Which is very different from what you get by constraining Tr(s_ii)/3.

So yes, we can write a FunctionMap to scale the value of sig_zz.
But this won't be convenient and understandable for the non-advanced easi user.
I m not very satisfied with this loss of backward compatibility, but it seems like the least worse solution.
Maybe we could go even further and remove 'effectiveConfiningStress' from the parameter list.

Comment (Thomas-Ulrich) [reply include/easi/component/OptimalStress.h:112]: The question is do we want to write that every time we use the module:
```    
    !EvalModel
    parameters: [b_xx,b_yy,b_zz,b_xy, b_xz, b_yz]
    model: !ConstantMap
        map:
                mu_d: a
        components: !OptimalStress
              constants:
                mu_s:      b
                strike:    c
                dip:       d
                rake:     90.0
                cohesion:  e
                s2ratio:   f
                R:         g
                effectiveConfiningStress: h
    components: !FunctionMap
        map:
           b_xx: return b_xx* (b_xx+b_yy+b_zz)/(3 *b_zz);
           b_yy: return b_yy* (b_xx+b_yy+b_zz)/(3 *b_zz);
           b_zz: return (b_xx+b_yy+b_zz)/3;
           b_xy: return b_xy* (b_xx+b_yy+b_zz)/(3 *b_zz);
           b_xz: return b_xz *(b_xx+b_yy+b_zz)/(3 *b_zz);
           b_yz: return b_yz* (b_xx+b_yy+b_zz)/(3 *b_zz);
           ```

Comment (uphoffc) [reply include/easi/component/OptimalStress.h:112]: Sorry if I don't get it, but say you only specify only s_zz (and drop effectiveConfiningStress). Could you recover the behaviour of effectiveConfiningStress with a FunctionMap? If yes, then I'd make s_zz the default and drop effectiveConfiningStress.

**Pull request closed without merging**